tachyon/crypto/sumcheck/sumcheck_verifier.h

```diff
@@ -0,0 +1,210 @@
+// Copyright 2022 arkworks contributors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.arkworks and the LICENCE-APACHE.arkworks
+// file.
+
+#ifndef TACHYON_CRYPTO_SUMCHECK_SUMCHECK_VERIFIER_H_
+#define TACHYON_CRYPTO_SUMCHECK_SUMCHECK_VERIFIER_H_
+
+#include <vector>
+
+#include "tachyon/crypto/sumcheck/sumcheck_prover_msg.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifier_msg.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifying_key.h"
+
+namespace tachyon::crypto {
+
+template <typename F>
+F InterpolateUniPoly(const std::vector<F>& poly, const F& evaluation_point);
+
+// Subclaim created when verifier is convinced
+template <typename F>
+struct Subclaim {
+  // The multi-dimensional point this multilinear evaluation is evaluated at
+  std::vector<F> point;
+  // The expected evaluation
+  F expected_evaluation;
+};
+
+template <typename MLE>
+class Verifier {
+ public:
+  using F = typename MLE::Field;
+
+  Verifier() = default;
+  explicit Verifier(const VerifyingKey& key)
+      : num_variables_(key.num_variables),
+        max_evaluations_(key.max_evaluations) {
+    randomness_.reserve(key.num_variables);
+    polynomials_received_.reserve(key.num_variables);
+  }
+
+  // Runs a "round" given a |ProverMsg|, sampling and storing a random value.
+  //
+  // While a true verifier round should perform actual verification, |Round()|
+  // merely samples and stores random values. Verifications will be performed
+  // altogether in |CheckAndGenerateSubclaim| after all prover and verifier
+  // sub-rounds are subsequently finished.
+  const VerifierMsg<F> Round(ProverMsg<F>& prover_msg) {
+    CHECK(!finished_);
+    F random_value = F::Random();
+    randomness_.push_back(random_value);
+    polynomials_received_.push_back(prover_msg.evaluations);
+
+    if (round_ == num_variables_) {
+      finished_ = true;
+    } else {
+      round_++;
+    }
+    return {random_value};
```

Comment:
```suggestion
    return {std::move(random_value)};
```
You'll have to add move constructor for `VerifierMsg`.